python/kotekan/testing.py

```diff
@@ -0,0 +1,504 @@
+                if self._next_check[r] <= 0:
+                    self.validated_fpga_seqs[r] = self.validated_fpga_seqs[r].union(
+                        self._check(r)
+                    )
+                    if len(self.validated_fpga_seqs[r]) >= self.len_test[r]:
+                        logger.info("Reader {} done.".format(r))
+                        active_readers.remove(r)
+                    self._next_check[r] = self.update_interval[r]
+
+    def _check(self, r):
+        logger.info("Updating reader {}.".format(r))
+        reader = self._readers[r]
+        time0 = time.time()
+        visraw = reader.update()
+        self.update_time[r].append(time.time() - time0)
+
+        self._validate_time(visraw, r)
+        validate(visraw, self.config, self.test_pattern)
+
+        return set(np.unique(visraw.time)[:]["fpga_count"])
+
+    def _validate_age(self, visraw, r):
+        """Check if the difference between current time and the frame timestamps is acceptable.
+
+        Parameters
+        ----------
+        visraw : VisRaw
+            The data.
+        r : int
+            ID of the reader that gave us visraw.
+        """
+        now = time.time()
+        times_valid = visraw.time[visraw.valid_frames.astype(np.bool)]
+        assert_equal(visraw.time.shape[1], visraw.num_freq)
+        assert_equal(visraw.time.shape[0], self.view_sizes[r])
+
+        times = np.unique(times_valid)
+
+        # check first time slot
+        for timestamp in times:
+            # timestamp = times[t]
+            seconds = timestamp["ctime"]
+            fpga_seq = timestamp["fpga_count"]
+            if fpga_seq in self.validated_fpga_seqs[r]:
+                logger.debug(
+                    "Frames with fpga_seq={} already read before, skipping...".format(
+                        fpga_seq
+                    )
+                )
+                continue
+            ctime = time.ctime(seconds)
+            age = now - seconds
+            logger.info(
+                "Validating fpga_seq {} from {} (age {}s).".format(fpga_seq, ctime, age)
+            )
+
+            error = age - self.update_time[r][-1] - self.update_interval[r]
+            if error > 0:
+                msg = "Frames with fpga_seq={} read {:.3}s later than expected (age={:.3}, readerID={}).".format(
+                    fpga_seq, error, age, r
+                )
+                if (
+                    0 <= self.error_threshold < error
+                    and len(self.validated_fpga_seqs[r]) > self.view_sizes[r]
+                ):
+                    raise ValidationFailed(msg)
+                else:
+                    logger.info(msg)
+
+    def _validate_time(self, visraw, r):
+        """Validation of all things time."""
+        self._validate_age(visraw, r)
+        self._validate_missing(visraw, r)
+
+    def _validate_missing(self, visraw, r):
+        """Check for missing frames and if this was to be expected."""
```

Comment:
It would be good to have an explanation of what this test is trying to do. I'm actually not sure, and the description is pretty vague.